syncer/atxsync/syncer.go

```diff
@@ -0,0 +1,323 @@
+					log.ZContext(ctx),
+					publish.Field().Zap(),
+					zap.String("peer", peer.String()),
+					zap.Error(err),
+				)
+				continue
+			}
+			s.logger.Info("downloaded epoch info",
+				log.ZContext(ctx),
+				publish.Field().Zap(),
+				zap.String("peer", peer.String()),
+				zap.Int("atxs", len(epochData.AtxIDs)),
+			)
+			// adding hashes to fetcher is not useful as they overflow the cache and are not used
+			// so we switch to asking best peers immediately
+			update := make(map[types.ATXID]int, len(epochData.AtxIDs))
+			for _, atx := range epochData.AtxIDs {
+				update[atx] = 0
+			}
+			select {
+			case <-ctx.Done():
+				return nil
+			case updates <- epochUpdate{time: time.Now(), update: update}:
+			}
+			// after first success switch to requests after interval
+			interval = s.cfg.EpochInfoInterval
+		}
+	}
+}
+
+func (s *Syncer) downloadAtxs(
+	ctx context.Context,
+	publish types.EpochID,
+	downloadUntil time.Time,
+	state map[types.ATXID]int,
+	updates chan epochUpdate,
+) error {
+	batch := make([]types.ATXID, 0, s.cfg.AtxsBatch)
+	batch = batch[:0]
+	var (
+		downloaded           = map[types.ATXID]bool{}
+		previouslyDownloaded = 0
+		start                = time.Now()
+		lastSuccess          time.Time
+		progressTimestamp    = start
+		nothingToDownload    = len(state) == 0
+	)
+
+	for {
+		// waiting for update if there is nothing to download
+		if nothingToDownload && s.closeToTheEpoch(publish, lastSuccess, downloadUntil) {
+			s.logger.Info(
+				"atx sync terminated",
+				log.ZContext(ctx),
+				publish.Field().Zap(),
+				zap.Int("downloaded", len(downloaded)),
+				zap.Int("total", len(state)),
+				zap.Int("unavailable", len(state)-len(downloaded)),
+				zap.Duration("duration", time.Since(start)),
+			)
+			return nil
+		}
+		if nothingToDownload {
+			select {
+			case <-ctx.Done():
+				return nil
+			case update := <-updates:
+				lastSuccess = update.time
+				for atx, count := range update.update {
+					state[atx] = count
+				}
+			}
+		} else {
+			// otherwise check updates periodically but don't stop downloading
+			select {
+			case <-ctx.Done():
+				return nil
+			case update := <-updates:
+				lastSuccess = update.time
+				for atx, count := range update.update {
+					state[atx] = count
+				}
+			default:
+			}
+		}
+
+		for atx, requests := range state {
+			if downloaded[atx] {
+				continue
+			}
+			exists, err := atxs.Has(s.db, atx)
+			if err != nil {
+				return err
+			}
+			if exists {
+				downloaded[atx] = true
+				continue
+			}
+			if requests >= s.cfg.RequestsLimit {
+				delete(state, atx)
+				continue
+			}
+			batch = append(batch, atx)
+			if len(batch) == cap(batch) {
+				break
+			}
+		}
+		nothingToDownload = len(batch) == 0
+
+		// report progress every 10% or every 20 minutes
+		if progress := float64(len(downloaded) - previouslyDownloaded); progress/float64(len(state)) > 0.1 ||
+			time.Since(progressTimestamp) > 20*time.Minute {
```

Comment:
Shouldn't the coefficient and this duration be configurable?

Reply:
i can make it configurable, but i felt like reasonable defaults for reporting is sufficient